### PR TITLE
Use PluginFormInterface on media types

### PR DESCRIPTION
**Pull request by paranojik**
https://www.drupal.org/node/2626396

### CODE REVIEW

File: src/MediaBundleForm.php [hunk 1]
```diff
@@ -92,7 +99,11 @@ public function form(array $form, FormStateInterface $form_state) {
           ),
         ),
       );
-      $form['type_configuration'][$plugin] += \Drupal::service('plugin.manager.media_entity.type')->createInstance($plugin, $plugin_configuration)->settingsForm($this->entity);
+      /** @var \Drupal\media_entity\MediaTypeBase $instance */
+      $instance = \Drupal::service('plugin.manager.media_entity.type')->createInstance($plugin, $plugin_configuration);
```
Comment (primsi) [inline src/MediaBundleForm.php]: Can we inject this?

File: src/MediaBundleForm.php [hunk 2]
```diff
@@ -18,6 +20,38 @@
...
+   * @var \Drupal\Core\Plugin\PluginFormInterface[]
+   */
+  protected $configurableInstances = [];
+
+  /**
+   * Manager for media entity type plugins.
+   *
+   * @var \Drupal\media_entity\MediaTypeManager
+   */
+  protected $media_type_manager;
```
Comment (slashrsm) [inline src/MediaBundleForm.php]: class member variables usually use cameCase.

File: src/MediaBundleForm.php [hunk 3]
```diff
@@ -92,7 +126,11 @@ public function form(array $form, FormStateInterface $form_state) {
           ),
         ),
       );
-      $form['type_configuration'][$plugin] += \Drupal::service('plugin.manager.media_entity.type')->createInstance($plugin, $plugin_configuration)->settingsForm($this->entity);
```
Comment (slashrsm) [inline src/MediaBundleForm.php:95]: Another \Drupal::service('plugin.manager.media_entity.type') call above that could leverage injected dependency now.

File: src/MediaBundleForm.php [hunk 4]
```diff
@@ -101,6 +139,28 @@ public function form(array $form, FormStateInterface $form_state) {
   /**
    * {@inheritdoc}
    */
+  public function validateForm(array &$form, FormStateInterface $form_state) {
+    parent::validateForm($form, $form_state);
+
+    // Let the selected plugin validate its settings.
+    $plugin = $this->entity->getType()->getPluginId();
```
Comment (slashrsm) [inline src/MediaBundleForm.php:146]: Plugin could have been changed at the submit. In that case this line won't get correct plugin ID any more. We should read from submitted values instead.

File: src/MediaTypeBase.php [hunk 5]
```diff
@@ -119,4 +120,14 @@ public function attachConstraints(MediaInterface $media) {}
   public function calculateDependencies() {
     return [];
   }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function validateConfigurationForm(array &$form, FormStateInterface $form_state) { }
```
Comment (slashrsm) [inline src/MediaTypeBase.php:134]: Should we also add default implementation for buildConfigurationForm()? There might be plugins that require no configuration. It is very unlikely though...

File: src/MediaBundleForm.php (hunk 4 above)
Comment (paranojik) [inline src/MediaBundleForm.php:146]: After discussing this we found that $this->entity already contains the newly built entity...so no need to change this.

**Comment by slashrsm** [comment]
Merged. Added few small fixed on merge: https://github.com/drupal-media/media_entity/commit/c51ec17b3da4b62e9fad72b472ff62a2da2bd5dc

**Pull request merged**